### PR TITLE
Moving to new version logic in daml-assistant.

### PR DESCRIPTION
**Pull request by fmota-da**
Moving toward the SDK version logic decided in #283.

- updating the release tarball version
- use SemVer to represent SDK versions (a bit of a hackjob, needs refactoring)
- project config definitively named daml.yaml as decided by @moritzkiefer-da  :)

### CODE REVIEW

File: daml-assistant/BUILD.bazel [hunk 1]
```diff
@@ -42,7 +44,9 @@ da_haskell_library(
         "directory",
         "extra",
         "filepath",
+        'microlens',
```
Comment (fmota-da) [inline daml-assistant/BUILD.bazel]:
```suggestion
        "microlens",
```

File: daml-assistant/daml-project-config/DAML/Project/Consts.hs [hunk 2]
```diff
@@ -59,7 +59,7 @@ damlConfigName = "config.yaml"
 
 -- | File name of config file in DAML_PROJECT (the project path).
 projectConfigName :: FilePath
-projectConfigName = "da.yaml"
+projectConfigName = "daml.yaml"
```
Comment (moritzkiefer-da) [inline daml-assistant/daml-project-config/DAML/Project/Consts.hs]: :tada:

File: daml-assistant/daml-project-config/DAML/Project/Types.hs [hunk 3]
```diff
@@ -44,27 +46,32 @@ newtype ProjectConfig = ProjectConfig
...
+    parseJSON y = do
+        verE <- V.fromText <$> Y.parseJSON y
+        case verE of
+            Left e -> fail ("Invalid SDK version: " <> e)
+            Right v -> pure (SdkVersion v)
+
+data SdkChannel
+    = Stable
+    | Unstable
+    | Custom [V.Identifier]
```
Comment (moritzkiefer-da) [inline daml-assistant/daml-project-config/DAML/Project/Types.hs]: Do you have any plan to use this?

File: daml-assistant/daml-project-config/DAML/Project/Types.hs [hunk 4]
```diff
@@ -44,27 +46,32 @@ newtype ProjectConfig = ProjectConfig
...
+    deriving (Eq, Ord, Show)
+
+versionChannel :: SdkVersion -> SdkChannel
+versionChannel (SdkVersion v) =
+    let ch = v L.^. V.release in
+    fromMaybe (Custom ch) $ lookup (map Just ch)
```
Comment (moritzkiefer-da) [inline daml-assistant/daml-project-config/DAML/Project/Types.hs]:
```suggestion
    case ch of
        [] -> Stable
        [x] | Just x == V.textual "unstable" -> Unstable
        _ -> Custom ch
```

Your choice but I find the logic here quite hard to parse.

File: daml-assistant/src/DAML/Assistant/Env.hs [hunk 5]
```diff
@@ -104,7 +105,11 @@ getSdk damlPath projectPathM =
     wrapErr "Determining SDK version and path." $ do
 
         sdkVersion <- firstJustM id
-            [ fmap (SdkVersion . pack) <$> lookupEnv sdkVersionEnvVar
+            [ lookupEnv sdkVersionEnvVar >>= \ vstrM -> pure $ do
+                vstr <- vstrM
+                v <- eitherToMaybe (V.fromText (pack vstr))
+                pure (SdkVersion v)
```
Comment (moritzkiefer-da) [inline daml-assistant/src/DAML/Assistant/Env.hs]:
```suggestion
                Just (SdkVersion v)
```
Given that there is another use of `pure` 3 lines above this for a different instance, I think it’s easier to understand this if you make it explicit which instance is being used.

File: daml-assistant/src/DAML/Assistant/Env.hs [hunk 6]
```diff
@@ -146,11 +149,10 @@ getLatestInstalledSdkVersion (DamlPath path) = do
             Left _ -> pure Nothing
             Right dirlist -> do
                 subdirs <- filterM (doesDirectoryExist . (dpath </>)) dirlist
-                -- TODO (FAFM): warn if subdirs /= dirlist
-                --  (i.e. $DAML_HOME/sdk is polluted with non-dirs).
-                let versions = filter ("nightly-" `isPrefixOf`) subdirs
-                    -- TODO (FAFM): configurable channels
-                pure $ fmap (SdkVersion . pack) (maximumMay versions)
+                let preversions = mapMaybe (eitherToMaybe . V.fromText . pack) subdirs
```
Comment (moritzkiefer-da) [inline daml-assistant/src/DAML/Assistant/Env.hs]: Should we throw an for subdirs that cannot be parsed instead of silently filtering them?
What does `preversions` stand for? It sounds similar to `prereleases` but that doesn’t seem to be accurate?

File: daml-assistant/src/DAML/Assistant/Install.hs [hunk 7]
```diff
@@ -39,16 +33,20 @@ import System.Posix.Files -- forget windows for now
...
+import qualified Data.SemVer as V
+import qualified Lens.Micro as L
 
 displayInstallTarget :: InstallTarget -> Text
 displayInstallTarget = \case
-    InstallChannel (SdkChannel c) -> "channel " <> c
-    InstallVersion (SdkVersion v) -> "version " <> v
+    InstallChannel Stable -> "channel stable"
+    InstallChannel Unstable -> "channel unstable"
+    InstallChannel ch -> "channel " <> pack (show ch)
```
Comment (moritzkiefer-da) [inline daml-assistant/src/DAML/Assistant/Install.hs]:
```suggestion
    InstallChannel ch@(Custom _) -> "channel " <> pack (show ch)
```
Worh being explicit imho so we get a warning if we add more constructors (you could also consider changing the formatting but that’s probably not important).

File: release/BUILD.bazel [hunk 8]
```diff
@@ -75,7 +75,7 @@ genrule(
     srcs = [
         ":config.yaml.tmpl",
         ":install.sh",
-        "//:component-version",
+        "//:VERSION",
```
Comment (moritzkiefer-da) [inline release/BUILD.bazel]: I think this change might break things as the latest version will probably be the latest version that has the +100 bump.

File: daml-assistant/BUILD.bazel [hunk 9]
```diff
@@ -14,10 +14,12 @@ deps = [
     'filepath',
     'http-conduit',
     'main-tester',
+    'microlens',
```
Comment (moritzkiefer-da) [inline daml-assistant/BUILD.bazel]: This is exactly one of the cases that I mentioned yesterday where you already have `lens` in your transitive dependency closure anyway (in this case, `da-hs-base` depends on it) so by adding `microlens` you potentially increase our dependency footprint instead of decreasing it.

File: daml-assistant/daml-project-config/DAML/Project/Types.hs (hunk 4 above)
Comment (fmota-da) [reply daml-assistant/daml-project-config/DAML/Project/Types.hs]: this is much nicer, thank you :-)

File: daml-assistant/src/DAML/Assistant/Env.hs (hunk 5 above)
Comment (fmota-da) [reply daml-assistant/src/DAML/Assistant/Env.hs]: yep, this makes sense

File: daml-assistant/src/DAML/Assistant/Env.hs (hunk 6 above)
Comment (fmota-da) [reply daml-assistant/src/DAML/Assistant/Env.hs]: it's just a variable name -_- I guess `semvers` would be more accurate

File: daml-assistant/src/DAML/Assistant/Install.hs (hunk 7 above)
Comment (fmota-da) [reply daml-assistant/src/DAML/Assistant/Install.hs]: :+1:

File: release/BUILD.bazel (hunk 8 above)
Comment (fmota-da) [reply release/BUILD.bazel]: The previous versions have "nightly-" which should make them get filtered out by the assistant's logic. Do you think this will mess things up with bintray? It's ok for the bintray version to still be +100 (we account for that in the URL logic above). It's also ok IMO for bintray downloads to break since we're moving to github releases with @garyverhaegen-da's help.

File: daml-assistant/BUILD.bazel (hunk 9 above)
Comment (fmota-da) [reply daml-assistant/BUILD.bazel]: So I should just depend on `lens`?

Comment (moritzkiefer-da) [reply daml-assistant/BUILD.bazel]: Either that or don’t use lenses.

File: release/BUILD.bazel (hunk 8 above)
Comment (moritzkiefer-da) [reply release/BUILD.bazel]: I am slightly worried that the `latest` URL will resolve to the `+100` version. If you are confident that this will not be a problem that’s fine :+1:

File: daml-assistant/BUILD.bazel (hunk 9 above)
Comment (fmota-da) [reply daml-assistant/BUILD.bazel]: Sadly the SemVer library only exposes lenses 😬

File: release/BUILD.bazel (hunk 8 above)
Comment (fmota-da) [reply release/BUILD.bazel]: It should resolve to the +100 version! I guess we'll find out if it breaks bintray downloads ... oh well.

File: daml-assistant/daml-project-config/DAML/Project/Types.hs (hunk 3 above)
Comment (fmota-da) [reply daml-assistant/daml-project-config/DAML/Project/Types.hs]: I might take this out. The concept of channel is premature.

File: daml-assistant/BUILD.bazel [hunk 10]
```diff
@@ -13,11 +13,13 @@ deps = [
     'extra',
     'filepath',
     'http-conduit',
+    'lens',
     'main-tester',
     'optparse-applicative',
     'process',
     'safe',
     'safe-exceptions',
+    "semver",
```
Comment (fmota-da) [inline daml-assistant/BUILD.bazel]:
```suggestion
    'semver',
```
